File: leetcode/two pointer/1358. Number of Substrings Containing All Three Characters/cpp/main.cpp
class Solution {
public:
    int numberOfSubstrings(string s) {
        int map[3] = {0, 0, 0};
        int start = 0;
        int end = 0;
        int freq = 0;
        int res = 0;
        int cnt = 0;
        
        while (end < s.size()) {
            if (map[s[end] - 'a'] == 0) {
                freq++;
            }
            
            map[s[end++] - 'a']++;
            
            while (freq == 3) {
                if (map[s[start] - 'a'] == 1) {
                    freq--;
                }
                map[s[start++] - 'a']--;
                cnt++;
            }
            
            res += cnt;
        }
        
        return res;
    }
};

File: leetcode/two pointer/1358. Number of Substrings Containing All Three Characters/js/main.js
// Time O(N)
// Space O(1)
const numberOfSubstrings = s => {
  let n = s.length;

  let start = 0;
  let end = 0;
  let k = 3;
  let map = { a: 1, b: 1, c: 1 };
  let res = 0;

  while (end < n) {
    if (map[s[end]] == 1) {
      k--;
    }

    map[s[end++]]--;

    while (k == 0) {
      if (map[s[start]] == 0) {
        k++;
      }

      res += n - end + 1;

      map[s[start++]]++;
    }
  }

  return res;
};
